### DISCUSSION TITLE
Intel 8265 Wifi on Nano

### DISCUSSION THREAD

**Discussion by tirankas** [OE4T/meta-tegra#716] [General]
Hello,
Trying to understand why the wlan0 interface takes more than 2 minutes to start 
(from power up).
Powerup log shows this:
[    7.556198] Loading modules backported from iwlwifi
[    7.587763] iwlwifi-stack-public:release/core52:8470:f4c7b3b5
[    7.997119] iwlwifi 0000:01:00.0: enabling device (0000 -> 0002)
[    8.004670] iwlwifi 0000:01:00.0: Direct firmware load for iwl-dbg-cfg.ini failed with error -2
[    8.013593] iwlwifi 0000:01:00.0: Falling back to user helper
[   68.761370] iwlwifi 0000:01:00.0: loaded firmware version 36.ad812ee0.0 8265-36.ucode op_mode iwlmvm
[   68.771427] iwlwifi 0000:01:00.0: Direct firmware load for iwl-debug-yoyo.bin failed with error -2
[   68.781083] iwlwifi 0000:01:00.0: Falling back to user helper
[  130.236751] iwlwifi 0000:01:00.0: Detected Intel(R) Dual Band Wireless AC 8265, REV=0x230
[  130.314184] iwlwifi 0000:01:00.0: base HW address: 14:f6:d8:75:94:cd
[  130.407561] thermal thermal_zone6: Registering thermal zone thermal_zone6 for type iwlwifi_1

This happens on the Nano.

wlan interface is shown only after 130 seconds...

I see people are publishing similar log but no one complains on the time it takes to identify the wlan card.

Thanks

**Comment by madisongh** [comment]
I don't know for sure, but you might try bundling its firmware into the initramfs, so it's present when the kernel starts (assuming you're using an initramfs in your setup, which is our default).   There's a `TEGRA_INITRD_INSTALL`  variable you can set to include the name of the relevant firmware package - the `tegra-minimal-initramfs` recipe will include that.

**Comment by tirankas** [comment]
Hi Matt,
I followed your suggestion:
TEGRA_INITRD_INSTALL += "linux-firmware-iwlwifi-8265"

I dont see a difference in terms of driver loading time, starts nearly at the same time.
I see this happens only after real root is mounted but this is mainly because the driver is loaded from the real root and not from the initramfs.

I think two options here:
1. Statically include the backport-iwlwifi into the kernel
2. Include the backported driver in the initramfs.

Not sure which is better.

Thanks

**Comment by tirankas** [comment]
@madisongh By the way, I looked at the initram cpio and I dont see the linux-firmware-iwlwifi-8265 in the /lib/firmware.
Am I doing something wrong?

**Reply by elPrac** [reply]
You will need to add this kernel config options:
CONFIG_EXTRA_FIRMWARE="iwlwifi-8265-22.ucode"
CONFIG_EXTRA_FIRMWARE_DIR="/lib/firmware"

And regarding driver loading time i was having the same problem, you can try my patch and look if it works for you 

```
From e8a4a907d482f82e08b09247d9e0cc9003759a9c Mon Sep 17 00:00:00 2001                                    
Date: Mon, 14 Sep 2020 21:12:47 -0500
Subject: [PATCH] Fix-iwlifi-8265-firmware-version

---
 drivers/net/wireless/intel/iwlwifi/iwl-8000.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/intel/iwlwifi/iwl-8000.c b/drivers/net/wireless/intel/iwlwifi/iwl-8000.c
index 20d08ddb4388..76fde0579e2e 100644
--- a/drivers/net/wireless/intel/iwlwifi/iwl-8000.c
+++ b/drivers/net/wireless/intel/iwlwifi/iwl-8000.c
@@ -71,11 +71,11 @@

 /* Highest firmware API version supported */
 #define IWL8000_UCODE_API_MAX  26
-#define IWL8265_UCODE_API_MAX  26
+#define IWL8265_UCODE_API_MAX  22

 /* Lowest firmware API version supported */
 #define IWL8000_UCODE_API_MIN  22
-#define IWL8265_UCODE_API_MIN  22
+#define IWL8265_UCODE_API_MIN  21

 /* NVM versions */
 #define IWL8000_NVM_VERSION        0x0a1d
-- 
2.17.1
```

**Reply by elPrac** [reply]
also you can find more clues following this [discussion](https://forums.gentoo.org/viewtopic-t-1001638-start-0.html)

**Reply by tirankas** [reply]
Thanks @elPrac , I am trying this now.

Still, would like to add the firmware file to initramfs and kernel driver as static, this would, in addition to @elPrac suggestion, make this much faster.
@madisongh Do you know what is wrong with this settings:

TEGRA_INITRD_INSTALL += "linux-firmware-iwlwifi-8265"

Thanks

**Reply by tirankas** [reply]
Hi @elPrac 
Tried this and still same issue:
[    8.083321] Loading modules backported from iwlwifi
[    8.089806] iwlwifi-stack-public:release/core52:8470:f4c7b3b5
[    8.399955] iwlwifi 0000:01:00.0: enabling device (0000 -> 0002)
[    8.409135] iwlwifi 0000:01:00.0: Direct firmware load for iwl-dbg-cfg.ini failed with error -2
[    8.418422] iwlwifi 0000:01:00.0: Falling back to user helper
[   68.764260] iwlwifi 0000:01:00.0: loaded firmware version 36.ad812ee0.0 8265-36.ucode op_mode iwlmvm
[   68.777680] iwlwifi 0000:01:00.0: Direct firmware load for iwl-debug-yoyo.bin failed with error -2
[   68.786726] iwlwifi 0000:01:00.0: Falling back to user helper
[  130.226838] iwlwifi 0000:01:00.0: Detected Intel(R) Dual Band Wireless AC 8265, REV=0x230
[  130.298095] iwlwifi 0000:01:00.0: base HW address: 34:7d:f6:a1:6e:27
[  130.391919] thermal thermal_zone6: Registering thermal zone thermal_zone6 for type iwlwifi_1

Checking the fw version file in debugfs:
FW prefix: iwlwifi-8265-
FW: release/core33::ad812ee0
Device: Intel(R) Dual Band Wireless AC 8265
Bus: pci

Dont understand why its picking up firmware version 36 after the patch...

Thanks

**Comment by elPrac** [comment]
In the previous log you post, it is also trying to load the same version if I'm not wrong the way it chooses a version is from a macro, you will probably need to look at the driver. I used 22.x version back then...
> [ 7.556198] Loading modules backported from iwlwifi
> [ 7.587763] iwlwifi-stack-public:release/core52:8470:f4c7b3b5
> [ 7.997119] iwlwifi 0000:01:00.0: enabling device (0000 -> 0002)
> [ 8.004670] iwlwifi 0000:01:00.0: Direct firmware load for iwl-dbg-cfg.ini failed with error -2
> [ 8.013593] iwlwifi 0000:01:00.0: Falling back to user helper
> [ 68.761370] iwlwifi 0000:01:00.0: loaded firmware version 36.ad812ee0.0 8265-36.ucode op_mode iwlmvm
> [ 68.771427] iwlwifi 0000:01:00.0: Direct firmware load for iwl-debug-yoyo.bin failed with error -2
> [ 68.781083] iwlwifi 0000:01:00.0: Falling back to user helper
> [ 130.236751] iwlwifi 0000:01:00.0: Detected Intel(R) Dual Band Wireless AC 8265, REV=0x230
> [ 130.314184] iwlwifi 0000:01:00.0: base HW address: 14:f6:d8:75:94:cd
> [ 130.407561] thermal thermal_zone6: Registering thermal zone thermal_zone6 for type iwlwifi_1

**Reply by tirankas** [reply]
Hi @elPrac 
The logic in the driver is simply listing of a table of cards with their info alo g with fw version. The numbers I put are correct but I don't understand the correlation.
I put max as 22 and min as 21.
From the log I see, it picked up 36...
I don't understand why 36 was chosen...
Maybe this is after going through all options, searching the fw files directly?
Maybe better I put 36 in the table...

Thanks

**Comment by tirankas** [comment]
I am still facing this issue, now it takes 60 seconds (as oppose to the 130 seconds before). Change was to using the version 33 of the backported iwlwifi which according to Intel is the last version that formally supports my 8265 card.

I am trying to change the iwlwifi driver to be static and not a module. However, build system forces it to be a module (saying conflict on the driver when building the defconfig during the config check phase). Is there a reason for not allowing me to build it statically into the kernel? Can I override the bsp settings by using my own .cfg?

Thanks

**Comment by madisongh** [comment]
> Is there a reason for not allowing me to build it statically into the kernel?

That's usually due to a dependency on another config setting that is set to `m`.

> Can I override the bsp settings by using my own .cfg?

Yes.

**Reply by tirankas** [reply]
My problem is probably not setti g the mac80211 and cfg80211 to yes. Will try to find the dependencies and mark them as yes.
I think the kernel config check phase should indicate the reason for not picking up the selected config.

**Reply by madisongh** [reply]
> I think the kernel config check phase should indicate the reason for not picking up the selected config.

I believe there are logs generated by the config check that do provide at least some of that information.
